### PR TITLE
add ci test for trl rewarding and ppo, fix backward failure in ppo ca…

### PR DESCRIPTION
**Pull request by sywangyi**
…used by rmsfusion

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
Traceback (most recent call last):
  File "/workspace/wangyi/optimum-habana/examples/trl/ppo.py", line 311, in <module>
    stats = ppo_trainer.step(question_tensors, response_tensors, rewards)
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/workspace/wangyi/optimum-habana/optimum/habana/trl/trainer/ppo_trainer.py", line 612, in step
    train_stats = self.train_minibatch(
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/workspace/wangyi/optimum-habana/optimum/habana/trl/trainer/ppo_trainer.py", line 844, in train_minibatch
    self.accelerator.backward(loss)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/accelerator.py", line 1966, in backward
    loss.backward(**kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_tensor.py", line 532, in backward
    torch.autograd.backward(
  File "/usr/local/lib/python3.10/dist-packages/torch/autograd/__init__.py", line 266, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/usr/local/lib/python3.10/dist-packages/torch/autograd/function.py", line 289, in apply
    return user_fn(self, *args)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpex/normalization/FusedRMSNorm.py", line 49, in backward
    inverse_root_mean_square, data_in, gamma = ctx.saved_tensors
RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [HPUBFloat16Type [1, 256, 4096]], which is output 0 of AddBackward0, is at version 64; expected version 63 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -58,10 +58,12 @@ def gaudi_llama_rmsnorm_forward(self, hidden_states):
         # mixed dtypes are not good for FusedRMSNorm, both inputs need to have same dtype
         if hidden_states.dtype != self.weight.dtype:
             orig_dtype = hidden_states.dtype
-            hidden_states = FusedRMSNorm.apply(hidden_states.to(self.weight.dtype), self.weight, self.variance_epsilon)
+            hidden_states = FusedRMSNorm.apply(
+                hidden_states.clone().to(self.weight.dtype), self.weight, self.variance_epsilon
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:62]: @sywangyi  why do we need clone here?

Comment (sywangyi) [reply optimum/habana/transformers/models/llama/modeling_llama.py:62]: we meet the similar issue with https://github.com/huggingface/optimum-habana/pull/507. I think. please see the history in 507

Comment (sywangyi) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: remove the clone, since ppo disable rms in ppo.py

**Comment by sywangyi** [comment]
hope it could be merged soon to avoid regression in trl.

**Comment by shepark** [comment]
I ran ci testing locally.
`GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_examples.py -s -v -k "ppo" 2>&1 | tee log_ppo.txt`
`GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_examples.py -s -v -k "reward_modeling" 2>&1 | tee log_reward.txt`

In both case, there's a failure.
For PPO,
___________________
tests/test_examples.py:619: in assert_no_regression
    assert_function(
E   AssertionError: 61.934252977371216 not less than or equal to 57.75 : for metric train_runtime. 
E   ===== Assessed metrics (measured vs thresholded baseline) =====
E   train_runtime: 61.934252977371216 vs 57.75
E   train_samples_per_second: 0.5166769350022155 vs 0.475
=========================== short test summary info ============================
FAILED tests/test_examples.py::MultiCardPPOExampleTester::test_ppo_llama-7b_multi_card
================= 1 failed, 48 deselected in 151.33s (0:02:31) =================

For REWARD,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_examples.py:619: in assert_no_regression
    assert_function(
E   AssertionError: 244.543 not less than or equal to 210.0 : for metric train_runtime. 
E   ===== Assessed metrics (measured vs thresholded baseline) =====
E   train_runtime: 244.543 vs 210.0
E   train_samples_per_second: 1.673 vs 1.9
=========================== short test summary info ============================
FAILED tests/test_examples.py::MultiCardRewardExampleTester::test_reward_modeling_llama-7b_multi_card
================= 1 failed, 48 deselected in 320.63s (0:05:20) =================

**Comment by sywangyi** [comment]
I test in my env. This two tests have both passed using your command, my SynapseAI is 1.16.2.

**Comment by shepark** [comment]
> I test in my env. This two tests have both passed using your command, my SynapseAI is 1.16.2.

Can you try it from 1.16.0-526? to see whether it's due to SyanpseAI version or any variation between systems.

**Comment by sywangyi** [comment]
also pass in 1.16.0-526

**Comment by shepark** [comment]
> also pass in 1.16.0-526

thank you for the confirmation.
can you recommend any safe numbers with considering minimum variations based on my data?

**Comment by sywangyi** [comment]
see if it works for you

**Comment by shepark** [comment]
> see if it works for you

thank you, it works.

**Review by shepark — APPROVED** [review]

**Review by regisss — APPROVED** [review]

**Pull request merged**